maxim/layers.py

```diff
@@ -1,65 +1,89 @@
 """
 Layers based on https://github.com/google-research/maxim/blob/main/maxim/models/maxim.py
+and reworked to cope with variable image dimensions
 """
 
-import einops
 import tensorflow as tf
 from tensorflow.experimental import numpy as tnp
-from tensorflow.keras import backend as K
 from tensorflow.keras import layers
 
 
 @tf.keras.utils.register_keras_serializable("maxim")
-class BlockImages(layers.Layer):
+class TFBlockImages(layers.Layer):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
-    def call(self, x, patch_size):
-        bs, h, w, num_channels = (
-            K.int_shape(x)[0],
-            K.int_shape(x)[1],
-            K.int_shape(x)[2],
-            K.int_shape(x)[3],
-        )
+    def call(self, image, patch_size):
+        bs, h, w, num_channels = (tf.shape(image)[0], tf.shape(image)[1], tf.shape(image)[2], tf.shape(image)[3])
+        ph, pw = patch_size
+        gh = h // ph
+        gw = w // pw
+        pad = [[0, 0], [0, 0]]
+        patches = tf.space_to_batch_nd(image, [ph, pw], pad)
+        patches = tf.split(patches, ph * pw, axis=0)
+        patches = tf.stack(patches, 3)  # (bs, h/p, h/p, p*p, 3)
+        patches_dim = tf.shape(patches)
+        patches = tf.reshape(patches, [patches_dim[0], patches_dim[1], patches_dim[2], -1])
+        patches = tf.reshape(patches, (patches_dim[0], patches_dim[1] * patches_dim[2], ph * pw, num_channels))
+        return [patches, gh, gw]
```

Comment:
What is the line number you're using for Black formatting? The line-numbers seem long and should be formatted accordingly.

Reply:
where I work we use 122, I am reformatting with 88.

Reply:
80 is the default. You can bump it to 90 (which is what I used).